### ISSUE TITLE
How can I filter socket hang up errors?

### ISSUE THREAD

**Issue by JSteunou** [nodejitsu/node-http-proxy#813]
I'm using node-http-proxy for a while now, and I think I'm using it quite right catching errors like this

```javascript
proxy.on('error', function(err, req, res) {
    // log
    // ...
    // handle
    if (!res.headersSent) {
        res.writeHead(500, { 'content-type': 'application/json' });
    }
    res.end(JSON.stringify({ error: 'proxy_error', reason: err.message }));
});
```

But I can have two cases that make a "socket hang up" error happen.

1. Crash behind
2. Client cancelling the request

Watching at my log I can make no differences between those two cases and it's quite frustrating. Before knowing about request abort (thank #527) I really though I have an error going on.

So, what's the best practice to filter the request.abort case out?

**Comment by theoephraim** [comment]
+1

**Comment by FoghostCn** [comment]
+1

**Comment by vvo** [comment]
You should be able to use the proxy error event or the error callback and check for res.socket.destroyed === true and error.code  === ECONNRESET which means the request was canceled

**Comment by JSteunou** [comment]
Merci @vvo ;) Will try that asap

**Comment by JSteunou** [comment]
It works very fine, thank you again @vvo 

@jcrugzz should I let this issue open? Maybe http-proxy should filter those error inside the library, or maybe it's better to let it unopinionated...

**Comment by vvo** [comment]
> or maybe it's better to let it unopinionated...

Yes it depends on the use case, sometime you will want the error to pop out if the client canceled the request.

**Comment by JSteunou** [comment]
What I thought. Closing it then.

**Comment by JSteunou** [comment]
Quick question again: I want to use [request abort](https://nodejs.org/docs/v0.12.8/api/http.html#http_request_abort) to prevent the underlying server to handle that client cancelled request. But, very surprisingly, `req.abort` is `undefined` O_O

Any thoughts?

**Comment by vvo** [comment]
nope sorry,

**Comment by JSteunou** [comment]
I think the `req` I got is the received request, not the request currently sent by http-proxy

**Comment by JSteunou** [comment]
Ok for the record, in caise anyone was interested at doing the same thing

### How to catch request client cancelled and abort ongoing proxied request 

```
proxy.on('proxyReq', function(proxyReq, req) {
    // keep a ref
    req._proxyReq = proxyReq;
});

proxy.on('error', function(err, req, res) {
  if (req.socket.destroyed && err.code === 'ECONNRESET') {
    req._proxyReq.abort();
  }
});
```

**Comment by jcrugzz** [comment]
This is a bug

**Comment by JSteunou** [comment]
Which one? Getting all errors even from client cancelled request or be forced to tricky save `proxyReq` to be able to `abort`?

**Comment by jcrugzz** [comment]
@JSteunou To put it simply, your workaround here should be handled within `http-proxy` for the `req.on('error')`

**Comment by JSteunou** [comment]
Would be nice :)

**Comment by jcrugzz** [comment]
@JSteunou Pull requests always welcome if you want to take a stab at it! :)

**Comment by jcrugzz** [comment]
also @vvo if you want the error handled, you could handle the error directly on `req.on('error')`. Im of the opinion that we need to cleanup our `proxyReq` regardless and if you want it as an error in your code the error can be handled there. Does this seem reasonable?

**Comment by vvo** [comment]
Yep makes sense

**Comment by JSteunou** [comment]
@jcrugzz I'm afraid I have the necessary solution and motivation but not the time to do it.

**Comment by jcrugzz** [comment]
@JSteunou it would be a simple change right around [here](https://github.com/nodejitsu/node-http-proxy/blob/master/lib/http-proxy/passes/web-incoming.js#L130-L132). You may be overestimating the time :). It would be a good opportunity for first contribution

**Comment by Lapixx** [comment]
Fairly old thread, but still an issue afaik. Resolving it by comparing the error code to `ECONNRESET` and ignoring the error as suggested before.

Would be happy to submit a PR to handle this exception internally, but aren't there any use cases where you'd actually want to handle this manually? At the very least you'd need to emit a new event instead (`abort` perhaps?). Any thoughts?

**Comment by cycold** [comment]
mark

**Comment by flevour** [comment]
Should we consider this closed?
